### PR TITLE
Feature/collision detection for merge

### PR DESCRIPTION
**Pull request by rttwente**
This is basically exactly the same PR as [this older PR from Rolf](https://github.com/RoboTeamTwente/roboteam_world/pull/72).
Unfortunately, due to the size and history of that older PR, it became nearly impossible to resolve the merge conflicts. To counter that:
1. branched off from that old PR, to not make the history of that branch even more complicated, and to keep it as a backup
2. Merged development into this new branch without merging new changes made to development
3. Re-implemented important changes that were ignored during this merge strategy

This branch is not fully tested, but time will prove the correctness of this PR, in combination with debugging.
Please Rolf, never again a PR this big :(

### CODE REVIEW

**Comment by rolfvdhulst** [comment]
I apologize for it being that big, but would like to point out it's as much a lot of other people's fault as it is mine. This PR relied on an other PR in utils which was open for 1.5 years... You cannot expect me to sit still and do nothing whilst I repeatedly ask for reviews.

**Comment by rolfvdhulst** [comment]
That being said, I'll run some checks on this PR to test if everything works as desired still.

File: observer/src/filters/vision/ball/BallAssigner.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  //this will probably become more complicated in the future as it will work on a per-object basis
+  for(const auto& observation : observations){
+      auto closest_prediction = std::min_element(assignment.op_pairs.begin(),assignment.op_pairs.end(),
+                                                 [&](const CameraGroundBallPredictionObservationPair& first,
+                                                     const CameraGroundBallPredictionObservationPair& second){
+                                                   return (observation.position - first.prediction.position).squaredNorm() <
+                                                       (observation.position - second.prediction.position).squaredNorm();
+                                                 });
+      //if there's no close ball, we will use it for new filters.
+      if(!assignment.op_pairs.empty() ||
```
Comment (alexander-de-ranitz) [inline observer/src/filters/vision/ball/BallAssigner.cpp]: If op_pairs is empty, closest_prediction will return an invalid iterator. closest_prediction->prediction.position-observation.position in L58 will therefore cause a segfault. Was the || supposed to be a &&?

Comment (rolfvdhulst) [reply observer/src/filters/vision/ball/BallAssigner.cpp]: Silly mistake: The not sign was not supposed to be there. Good catch.

**Comment by rolfvdhulst** [comment]
Fixed one more bug where the angle of robots would be 0 because I forgot to readd an offset I was substracting. 
For the rest everything seems fine from my side. I would highly suggest running with a real field to ensure that nothing crashes

**Comment by itsasimpleplan** [comment]
Hi! Floris asked me to have a look at this PR and observe what is going wrong. Made some screenshots and video's of the weird things I observed. For all other submodules in roboteam_suite I ran development. For world I ran the branch of this PR. 
I ran it with kick off us prepare and then kick off. 

Hereby already a list, screenshots and video's are bellow:
- Spinning robots (happens when everything starts up, before selecting a play)
- No yellow robots visible (when robots are spinning, before play selection)
- No robots to be seen (also when robots are spinning, before play selection)
- Weird paths drawn
- Robots chilling at the borders.

Spinning robots: 
https://user-images.githubusercontent.com/28604518/166208923-0e5f1992-d78b-4fc3-9a22-e729720fc208.mp4

No ryellow obots visible: interface_test

No robots:
Screenshot from 2022-05-02 10-54-10

Weird paths drawn:
Screenshot from 2022-04-25 10-59-57

Robots chilling:
sim_test

**Comment by rolfvdhulst** [comment]
@itsasimpleplan Thanks for the review, I could reproduce the issues and am working on them.

**Comment by rolfvdhulst** [comment]
I understand what is going wrong now, but it's a Grsim specific issue. Basically GrSIM is sending timestamps which start at t=0 whenever you start  grsim again instead of using the utc time. Our software was trying to extrapolate until now, which means that the estimated robot positions would be extrapolated 52 years into the future... This also explains why I didn't catch it earlier as I primarily tested using ER-Force's simulator.
For now I fixed it by disabling it, but I made an issue for GrSIM to fix this since this is actually bad behavior from their application. The 'old' version should actually work better with real robots as well.

**Comment by rolfvdhulst** [comment]
I fixed it and left it an option: in VisionFilter.cpp (timeextrapolation policy)
Ideally it would automatically switch depending on whether we are in the simulator or in real-life, but that would require me to listen to the interface/settings, which is currently not connected to observer, so I consider that as outside of the scope of this PR for now.

**Comment by rttwente** [comment]
Based on some quick testing, the fixes seem to work well for me, so thank you Rolf! I did notice that handling feedback is not yet implemented, @rolfvdhulst is this something you are already working on / that you would like to work on, or would it be better for one of us to take that up?

**Comment by rolfvdhulst** [comment]
@rttwente I would definitely like to work on feedback handling! It's not in here because this PR is very large, but I suspect a basic implementation should be quite straightforward.

**Comment by rolfvdhulst** [comment]
Oops, just wanted to comment but clicked the wrong button. 
I would only ensure that this PR is verified with real robots now by testing on a real field, and then it can be merged.

**Review by Florisheinen1 — APPROVED** [review]
LGTM lol

**Pull request merged**